oauthproxy.go

```diff
@@ -842,6 +844,13 @@ func (p *OAuthProxy) AuthOnly(rw http.ResponseWriter, req *http.Request) {
 
 // SkipAuthProxy proxies allowlisted requests and skips authentication
 func (p *OAuthProxy) SkipAuthProxy(rw http.ResponseWriter, req *http.Request) {
+	if p.addHeadersToSkipped {
```

Comment:
I think there's a security flaw with this design:

How do you differentiate between an allowlisted route where you opportunistically extract the session into the headers for the backend to use for authorization logic

VS

Requests that don't have a session, but hit the allowlisted route with an attacker injecting custom X-Forwarded-* headers spoofing other users.

I don't think there's a viable way to close this security hole. In general with authoritative headers we set that we intend backends to use for authorization logic we've been very paranoid & error for locked down security so no identities or groups inside of them could ever be spoofed by an attacker.

Reply:
A workaround that takes more work, but has no security holes:

If you control the backend code, if you receive the session cookie -- you can pass it to the `/oauth2/userinfo` endpoint yourself in a subrequest from your backend controller to extract the user/email/group details from the session.

Reply:
I don't believe this is actually the case - any headers we have configured to inject get stripped from the request before the session tries to inject them.  In this example I'm using the following "alpha" config:

```
injectRequestHeaders:
- name: Authorization
  values:
  - basicAuthPassword:
      value: cGFzc3dvcmQ=
    claim: user
    prefix: 'Basic '
- name: X-Forwarded-Groups
  values:
  - claim: groups
- name: X-Forwarded-User
  values:
  - claim: user
- name: X-Forwarded-Email
  values:
  - claim: email
- name: X-Forwarded-Preferred-Username
  values:
  - claim: preferred_username
- name: X-Forwarded-Access-Token
  values:
  - claim: access_token
```

and I've configured `/anything/skipped` as a skipped route.  I get the same result below both with and without `add_headers_to_skipped = true`:

```
$ curl -v -H 'X-Forwarded-For: 123.45.67.89' \
    -H 'X-Forwarded-Email: dangerous@attacker.com' \
    -H 'X-Forwarded-User: another@attacker.com' \
    -H 'Authorization: Basic dXNlcjpwYXNzd29yZA==' \
    -H 'X-Header-Test: testing' \
    -H 'X-Forwarded-Access-Token: sometoken' \
    http://localhost:4180/anything/skip
*   Trying ::1:4180...
* Connected to localhost (::1) port 4180 (#0)
> GET /anything/skip HTTP/1.1
> Host: localhost:4180
> User-Agent: curl/7.73.0
> Accept: */*
> X-Forwarded-For: 123.45.67.89
> X-Forwarded-Email: dangerous@attacker.com
> X-Forwarded-User: another@attacker.com
> Authorization: Basic dXNlcjpwYXNzd29yZA==
> X-Header-Test: testing
> X-Forwarded-Access-Token: sometoken
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Origin: *
< Content-Length: 361
< Content-Type: application/json
< Date: Wed, 24 Feb 2021 10:53:44 GMT
< Server: gunicorn/19.9.0
< 
{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip", 
    "Host": "localhost:4180", 
    "User-Agent": "curl/7.73.0", 
    "X-Header-Test": "testing"
  }, 
  "json": null, 
  "method": "GET", 
  "origin": "123.45.67.89, 172.23.0.1", 
  "url": "http://localhost:4180/anything/skip"
}
```

The sanity-check `X-Header-Test` gets through, but the forwarded email, user and access-token are stripped, as is the basic auth.

If I pass in a _valid_ JWT bearer token to simulate authenticating when it's optional:

```
$ curl -v -H 'Authorization: Bearer {valid token goes here}' \
    -H 'X-Forwarded-Email: dangerous@attacker.com' \
    -H 'X-Header-Test: testing' \
    http://localhost:4180/anything/skip
```

Then I get this with the default `add_headers_to_skipped = false`

```
{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip", 
    "Host": "localhost:4180", 
    "User-Agent": "curl/7.73.0", 
    "X-Header-Test": "testing"
  }, 
  "json": null, 
  "method": "GET", 
  "origin": "172.23.0.1", 
  "url": "http://localhost:4180/anything/skip"
}
```

and this with it set to true:

```
{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip", 
    "Authorization": "Basic {proper basic auth with the forwarded username}", 
    "Host": "localhost:4180", 
    "User-Agent": "curl/7.73.0", 
    "X-Forwarded-Access-Token": "{the token I authenticated with}", 
    "X-Forwarded-Email": "{correct value from token}", 
    "X-Forwarded-Preferred-Username": "{correct value from token}", 
    "X-Forwarded-User": "{correct value from token}", 
    "X-Header-Test": "testing"
  }, 
  "json": null, 
  "method": "GET", 
  "origin": "172.23.0.1", 
  "url": "http://localhost:4180/anything/skip"
}
```

In neither case does the `dangerous@attacker.com` header get through.

Reply:
Thanks for digging deeper! I forgot we recently moved the header squashing to the `headersChain` which gets called here now (I don't think that used to be the case for skip auth proxy).